plugins/pulp_docker/plugins/token_util.py

```diff
@@ -0,0 +1,92 @@
+from cStringIO import StringIO
+import json
+import logging
+import urllib
+import urlparse
+
+from nectar.request import DownloadRequest
+
+
+_logger = logging.getLogger(__name__)
+
+
+def add_auth_header(request, token):
+    """
+    Adds the token into the request's headers as specified in the Docker v2 API documentation.
+
+    https://docs.docker.com/registry/spec/auth/token/#using-the-bearer-token
+
+    :param request: a download request
+    :type  request: nectar.request.DownloadRequest
+    :param token: a Bearer token to be inserted into the Authorization header
+    :type  token: basestring
+    """
+    if request.headers is None:
+        request.headers = {}
+    request.headers['Authorization'] = 'Bearer %s' % token
+
+
+def request_token(downloader, request, response_headers):
+    """
+    Attempts to retrieve the correct token based on the 401 response header.
+
+    According to the Docker API v2 documentation, the token be retrieved by issuing a GET
+    request to the url specified by the `realm` within the `WWW-Authenticate` header. This
+    request should add the following query parameters:
+
+        service: the name of the service that hosts the desired resource
+        scope:   the specific resource and permissions requested
+
+    https://docs.docker.com/registry/spec/auth/token/#requesting-a-token
+
+    :param downloader: Nectar downloader that will be used to issue a download request
+    :type  downloader: nectar.downloaders.threaded.HTTPThreadedDownloader
+    :param request: a download request
+    :type  request: nectar.request.DownloadRequest
+    :param response_headers: headers from the 401 response
+    :type  response_headers: basestring
+    :return: Bearer token for requested resource
+    :rtype:  str
+    """
+    auth_info = parse_401_response_headers(response_headers)
+    try:
+        token_url = auth_info.pop('realm')
+    except KeyError:
+        raise IOError("No realm specified for token auth challenge.")
+
+    parse_result = urlparse.urlparse(token_url)
+    query_dict = urlparse.parse_qs(parse_result.query)
+    query_dict.update(auth_info)
+    url_pieces = list(parse_result)
+    url_pieces[4] = urllib.urlencode(query_dict)
```

Comment:
It's a pain in the ass, but for the sake of readability I'd rather see this either be spelled out, e.g. `url_pieces = [parse_result.scheme, parse_result.hostname, ...]`, or just throw a comment along the lines of `# replace the parsed query string, urlpieces[4]`